### PR TITLE
Client - Implemented message serialization

### PR DESCRIPTION
**Pull request by matjazv**
# Description of change

- implemented #296 
- fixed parsing token scheme in foundry output 

## Type of change

- Bugfix (a non-breaking change which fixes an issue)
- Enhancement (a non-breaking change which adds functionality)

## How the change has been tested

All current unit tests passed.

## Change checklist

Add an `x` to the boxes that are relevant to your changes, and delete any items that are not.

- [x] My code follows the contribution guidelines for this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [ ] I have added tests using CTest that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes

### CODE REVIEW

File: src/client/api/json_parser/payloads.c [hunk 1]
```diff
@@ -76,6 +85,36 @@ static cJSON* json_tx_essence_serialize(transaction_essence_t* es) {
...
+  switch (type) {
+    case CORE_MESSAGE_PAYLOAD_TAGGED:
+      if (json_tagged_deserialize(essence_payload, tagged_data) != 0) {
+        printf("[%s:%d] Can not deserialize tagged data\n", __func__, __LINE__);
+        return -1;
+      }
+      *payload_len = tagged_data_serialize_len(*tagged_data);
+      break;
+    default:
+      printf("[%s:%d] Unknown payload type\n", __func__, __LINE__);
```
Comment (oopsmonk) [inline src/client/api/json_parser/payloads.c]:
```suggestion
      printf("[%s:%d] Unsupported payload type\n", __func__, __LINE__);
```

File: src/client/api/json_parser/payloads.c [hunk 2]
```diff
@@ -182,5 +222,115 @@ cJSON* json_transaction_serialize(transaction_payload_t* tx) {
...
+      cJSON_Delete(tagged_data_payload);
+      return NULL;
+    }
+  }
+
+  return tagged_data_payload;
+}
+
+int json_tagged_deserialize(cJSON* payload, tagged_data_t** tagged_data) {
+  if (!payload) {
```
Comment (oopsmonk) [inline src/client/api/json_parser/payloads.c]: `tagged_data` should be checked too.
```suggestion
  if (!payload || *tagged_data) {
```

**Review by oopsmonk — CHANGES_REQUESTED** [review]

File: src/core/models/payloads/tagged_data.h [hunk 3]
```diff
@@ -17,6 +17,7 @@
  * The payload type of tagged data is 5
  */
 typedef struct {
+  uint32_t type;     ///< Set to value 5 to denote a Tagged Data payload.
```
Comment (oopsmonk) [inline src/core/models/payloads/tagged_data.h]: the `type` is a constant 5, why do we need it in this struct?
```suggestion

```

File: src/core/models/payloads/tagged_data.c [hunk 4]
```diff
@@ -240,21 +241,30 @@ void tagged_data_print(tagged_data_t *tagged_data, uint8_t indentation) {
...
     }
   } else {
-    printf("%sTag:\n", PRINT_INDENTATION(indentation));
+    printf("%s\tTag:\n", PRINT_INDENTATION(indentation));
   }
 
   // binary data
   if (tagged_data->data) {
-    printf("%sData: ", PRINT_INDENTATION(indentation));
-    dump_hex_str(tagged_data->data->data, tagged_data->data->len);
```
Comment (oopsmonk) [inline src/core/models/payloads/tagged_data.c:256]: what if data is binary, not string? we should represent the original data in hex string, the user can decode as needed.

File: src/client/api/json_parser/payloads.c (hunk 1 above)
Comment (matjazv) [reply src/client/api/json_parser/payloads.c]: Done.

File: src/client/api/json_parser/payloads.c (hunk 2 above)
Comment (matjazv) [reply src/client/api/json_parser/payloads.c]: Fixed. Thanks!

File: src/core/models/payloads/tagged_data.h (hunk 3 above)
Comment (matjazv) [reply src/core/models/payloads/tagged_data.h]: `Type` has been removed from a structure. Thanks.

File: src/client/api/json_parser/payloads.c (hunk 2 above)
Comment (oopsmonk) [reply src/client/api/json_parser/payloads.c]: https://github.com/iotaledger/iota.c/pull/305/commits/7527ace33618e491e9af74f462f24aa2e01c2204 doesn't fix the problem. 
the value `*tagged_data` should be checked as the suggestion above, not the address of the pointer `tagged_data`.
same as line [90](https://github.com/iotaledger/iota.c/pull/305/files#diff-f16bd3240fb36df6ddc7b455d53f2e34fbe0330fb3f54bd4a14d3cc1b8e00012R90), it should be 
```
if (!essence_payload || *tagged_data) {
```

Comment (matjazv) [reply src/client/api/json_parser/payloads.c]: Thank you for the explanation. I got it where it was a problem. It's fixed now.

**Review by oopsmonk — APPROVED** [review]

**Pull request merged**